tf5muxserver/mux_server_PrepareProviderConfig.go

```diff
@@ -42,16 +43,22 @@ func (s muxServer) PrepareProviderConfig(ctx context.Context, req *tfprotov5.Pre
 			resp.Diagnostics = append(resp.Diagnostics, res.Diagnostics...)
 		}
 
-		if res.PreparedConfig != nil {
-			// This could check equality to bypass the error, however
-			// DynamicValue does not implement Equals() and previous mux server
-			// implementations have not requested the enhancement.
-			if resp.PreparedConfig != nil {
-				return nil, fmt.Errorf("got a PrepareProviderConfig PreparedConfig response from multiple servers, not sure which to use")
-			}
+		// Do not check equality on missing PreparedConfig or unset PreparedConfig
+		if res.PreparedConfig == nil {
+			continue
+		}
 
-			resp.PreparedConfig = res.PreparedConfig
+		equal, err := dynamicValueEquals(schemaType(s.providerSchema), res.PreparedConfig, resp.PreparedConfig)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to compare PrepareProviderConfig PreparedConfig responses: %w", err)
 		}
+
+		if !equal {
+			return nil, fmt.Errorf("got different PrepareProviderConfig PreparedConfig response from multiple servers, not sure which to use")
+		}
```

Comment:
You are correct that upfront, the server creation will verify that all servers have the same schema, or otherwise return an error.

`PreparedConfig` is a data structure that's implemented from that schema and has an interesting history in Terraform. It allowed providers to modify the incoming configuration value of a provider and return a different result (e.g. marking a schema attribute as required, receiving no configuration value, then filling the configuration value via environment variable value, which Terraform would then "allow" in the required sense).

As with other parts of Terraform, this modification possibility was left for backwards compatibility (e.g. with terraform-plugin-sdk/v2), however going forward this capability is being discouraged in implementations. For example, terraform-plugin-framework uses the read-only `tfsdk.Config` type when handling this data. Since both terraform-plugin-sdk/v2 and terraform-plugin-framework do however _copy_ the configuration into `PreparedConfig`, we need to allow equal return values, but still want to prevent unequal values given our more recent implementation goals.

Reply:
OK so, I reached out to @bflad and he clarified things for me. What we are verifying here is that the configuration returned by the muxed servers, after `PrepareProviderConfig` (protocol version 5) and `ValidateProviderConfig` (protocol version 6), is STILL consistent.

While the HCL configuration fed to the provider is one and only one, because we are _muxing_, each server might be implementing transformations of the configuration, and the result would be incompatible.

This whole check is to avoid this: if one creates a provider muxing servers, and one (or more) of them end up creating different configurations, this library will detect it and return an error.